codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/RestJsonDocumentBodyGenerator.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.aws.typescript.codegen;
+
+import java.util.List;
+import java.util.Set;
+import java.util.TreeSet;
+import software.amazon.smithy.codegen.core.SymbolProvider;
+import software.amazon.smithy.model.knowledge.HttpBinding;
+import software.amazon.smithy.model.knowledge.NeighborProviderIndex;
+import software.amazon.smithy.model.neighbor.Walker;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeVisitor;
+import software.amazon.smithy.model.traits.JsonNameTrait;
+import software.amazon.smithy.model.traits.TimestampFormatTrait;
+import software.amazon.smithy.typescript.codegen.TypeScriptWriter;
+import software.amazon.smithy.typescript.codegen.integration.DocumentMemberDeserVisitor;
+import software.amazon.smithy.typescript.codegen.integration.DocumentMemberSerVisitor;
+import software.amazon.smithy.typescript.codegen.integration.HttpBindingProtocolGenerator;
+
+/**
+ * Handles general components across the AWS JSON protocols that have HTTP bindings.
+ * It handles reading and writing from document bodies, including generating any
+ * functions needed for performing serde.
+ *
+ * This builds on the foundations of the {@link HttpBindingProtocolGenerator} to handle
+ * components of binding to HTTP requests and responses.
+ *
+ * @see JsonShapeSerVisitor
+ * @see JsonShapeDeserVisitor
+ * @see JsonMemberSerVisitor
+ * @see DocumentMemberDeserVisitor
+ * @see <a href="https://awslabs.github.io/smithy/spec/http.html">Smithy HTTP protocol bindings.</a>
+ */
+abstract class RestJsonDocumentBodyGenerator extends HttpBindingProtocolGenerator {
```

Comment:
Why is this called RestJsonDocumentBodyGenerator? Seems like it generates the whole protocol, not just document bodies.